Issue #12: Add configuration to filter to allow or forbid specific tokens

bugfolder opened the pull request:
Fixes https://github.com/backdrop-contrib/token_filter/issues/12.

bugfolder committed:
Issue #12: Add configuration to filter to allow or forbid specific tokens.

bugfolder committed:
Update the UI a bit.

bugfolder committed:
Rename a variables: $part —> $name.

bugfolder committed:
Add warning about cache-clearing if token filter has changed.

bugfolder committed:
Change forbid/allow to ignore/evaluate.

bugfolder committed:
Only show the token tree once on the filter/tips page.

bugfolder committed:
Change “evaluate” to “replace” throughout module.

argiepiano commented on token_filter.install:
```diff
@@ -0,0 +1,24 @@
+    $config = config($config_name);
+    $data = $config->getData();
+    if (in_array('filter_tokens', array_keys($data['filters']))) {
+      $data['filters']['filter_tokens']['settings'] = array(
```
I'd suggest using a `+=` to preserve any previous settings in this array (for example, by a custom module). 

```suggestion
      $data['filters']['filter_tokens']['settings'] += array(
```

argiepiano commented on token_filter.install:
```diff
@@ -0,0 +1,24 @@
+ * Initialize the token filter settings for formats from previous versions.
+ */
+function token_filter_update_1000() {
+  $format_configs = config_get_names_with_prefix('filter.format.');
```
The `.` at the end of `filter.format.` is not needed but does no harm.

argiepiano replied:
Actually, I changed my mind. The period is a good idea. Otherwise this may match something like `filter.formatting.json` as well without the period.

argiepiano commented on token_filter.module:
```diff
@@ -64,9 +64,123 @@ function _token_filter_filter_tokens($text, $filter, $format, $langcode, $cache,
+    '#default_value' => isset($filter->settings['tokens']) ? $filter->settings['tokens'] : '',
+    '#description' => t('Enter tokens to ignore or replace, one per line, including the brackets. You can use "*" as a wildcard, for example, [user:*]. Select "Ignore" above and leave this blank to replace all recognized tokens. Ignored tokens will be left unaltered in the text.'),
+  );
+  return $settings;
```
Wondering if the token tree may be more helpful here, than in the function `_token_filter_filter_tips()`? You don't really need to see the tokens over there, but here, it'd be helpful to have a list. I realize the returned values of this function are shown in a modal window, but I have actually used a token tree inside a modal (in the module `token_block`) and it works. You get a second modal with the tree, which you can move to the side.

Also, wondering if it may be helpful to show the token tree for all core entity tokens here, since those are likely the ones you'll use. You would do this by passing the key 'token_types' => array('node', 'user', ...)  when you call `theme('token_tree'...)`. I see you are only passing `user`.

bugfolder replied:
I was only passing user because that was the only core entity that was guaranteed to be usable. There's not necessarily a node context available for `[node:?]` tokens to get rendered. Do you think it's better to list more tokens even if some of them may not be recognized, or to be conservative and leave out tokens that may not be recognized?

argiepiano replied:
To me, showing at least the `node` tokens is important. For example, `[node:author:mail]` could be a token I may want to forbid, and it'd be nice to see it in the tree.

~~BTW, I just found a bug generated by the new code I'll report in a second.~~

argiepiano commented on token_filter.module:
```diff
@@ -97,19 +211,50 @@ function _token_filter_get_entity_mapping($value_type = 'token', $value = NULL,
+    $output = t('Global tokens will be replaced with their respective token values (e.g. [site:name] or [current-page:title])!qualifier.', array('!qualifier' => $qualifier));
+    // Only show the token tree once on a page, even if the token filter is used
+    // in multiple text formats.
+    $token_tree = &backdrop_static(__FUNCTION__);
```
A static cache is not really needed here. The function that builds the tree (`token_build_tree()`) already uses a static cache. But it does not harm...

bugfolder replied:
The point of the static variable is not for caching, but that the token tree only gets rendered once on the `filter/tips` page, even if multiple formats use the Token Filter filter. If a site has, say, 3 different text formats that all use Token Filter, the user doesn't need to see three copies of the full token tree. (In fact, this usage is sort of an anti-cache: normally you cache so that you can use the thing more than once during a page load, but here we're using the static variable to make sure that it doesn't get used more than once.)

argiepiano replied:
Got it

bugfolder committed:
Update token_filter.install

Add support for custom module having already altered the Token Filter settings prior to this update.

Co-authored-by: Alejandro Cremaschi <argiepiano@yahoo.com>

argiepiano commented on token_filter.module:
```diff
@@ -64,9 +64,123 @@ function _token_filter_filter_tokens($text, $filter, $format, $langcode, $cache,
+    return $text;
+  }
+
+  $replace = $filter->settings['replace'];
```
~~This throws a `Notice: Undefined index: replace in _token_filter_filter_tokens() (line 74 of XXX/modules/token_filter/token_filter.module).`~~

To reproduce:
1. In the Configure filter modal, leave the Tokens textfield empty (no tokens forbidden)
2. "Update" and "Save configuration"
3. Create a node with some tokens in the body
4. Save. Clear the cache

See the notice.

argiepiano replied:
Actually, my report is completely wrong (too early in the morning!). I'm not able to reproduce this again. I'm not sure what happened

yorkshire-pudding replied:
@argiepiano @bugfolder 
I think I've reproduced something related to this.  Follow the steps above. Go to admin/content/node  Clear "Page and else" cache (think clearing all caches would also do it but narrowed down to this).

Got warnings:
> Warning: Undefined array key "evaluate" in _token_filter_filter_tokens() (line 74 of /app/docroot/modules/token_filter/token_filter.module).

> Warning: Undefined array key "tokens" in _token_filter_filter_tokens() (line 85 of /app/docroot/modules/token_filter/token_filter.module).

> Deprecated function: preg_quote(): Passing null to parameter #1 ($str) of type string is deprecated in preg_quote() (line 85 of /app/docroot/modules/token_filter/token_filter.module).

If I revert the patch then this error goes away (spent ages trying to figure this out in case a module I'm porting was causing it 🤦 )

argiepiano replied:
@yorkshire-pudding you may be working with an old or wrong version of token_filter. Line 74 in the dev version after applying the patch is a comment. Be sure to apply this patch to the dev version. The patch will not apply correctly to the latest release.

argiepiano replied:
Sorry, my bad. I was looking at the dev versino before applying the patch. Line 74 does exist.

However, I'm not able to reproduce. I've followed the steps.

yorkshire-pudding replied:
I've tried to reproduce on vanilla site but no joy. Must be something else interacting with it as that site has a number of modules I maintain or am porting or attempted to port so I'll have to work out which it is.

argiepiano replied:
Bingo! I was able to reproduce. 

The issue happens because, when you first enable the filter after a clean installation (that is, a virgin site that has never had Token Filter installed before - this is related to another minor bug I'll report in the issue), then enable he filter WITHOUT editing and saving the "Configure" modal =  then this is what you get in the config
```json
        "filter_tokens": {
            "status": 1,
            "weight": 0,
            "module": "token_filter",
            "settings": []
        }
```

As you can see, settings is empty, which throws the warnings. I think a solution might be to initialize with default settings. Normally this should look like this:

```json
        "filter_tokens": {
            "status": 1,
            "weight": 0,
            "settings": {
                "replace": "0",
                "tokens": ""
            },
            "module": "token_filter"
        }
```

yorkshire-pudding replied:
Confirmed that config did look like that and saving the config fixes it

yorkshire-pudding commented on token_filter.module:
```diff
@@ -11,8 +10,9 @@
-    'description' => t('The usage of this filter should be restricted to trusted users only as tokens with sensitive data could be exposed.'),
+    'description' => t('The usage of this filter might need to be restricted to trusted users only, as tokens with sensitive data could be exposed. You can also limit the tokens replaced by this filter in its configuration.'),
     'process callback' => '_token_filter_filter_tokens',
+    'settings callback' => '_token_filter_settings',
     'tips callback' => '_token_filter_filter_tips',
```
@bugfolder 
On the basis of @argiepiano 's great detective 🕵️ skills, here is a suggested change to the PR:

```suggestion
    'settings callback' => '_token_filter_settings',
    'default settings' => array(
      'replace' => 0,
      'tokens' => '',
    ),
    'tips callback' => '_token_filter_filter_tips',
```

bugfolder committed:
Update token_filter.module

Co-authored-by: Martin Price <martin@systemhorizons.co.uk>

yorkshire-pudding reviewed (approved):
Looks and works great. Thanks @bugfolder

argiepiano commented:
Wondering if this can be merged and a new release can be made?

bugfolder commented:
> Wondering if this can be merged and a new release can be made?

Both done. Thanks, @argiepiano and @yorkshire-pudding for input and reviews!
